### PR TITLE
Option to disable metadata fields check

### PR DESCRIPTION
**Pull request by rajaram5**
I try to implement this feature (https://dtl-fair.atlassian.net/secure/RapidBoard.jspa?rapidView=3&projectKey=FDP&view=detail&selectedIssue=FDP-327). I have introduced few overloading methods which gives an option to the client to enable or disable metadata fields check. I also introduced a new argument in the Methods of the MetadataUtils class to get the metadata RDF statements with different RDF model. We use DCAT based model has a default now but in future we can get RDF statements with different data model (Example: schema.org)

### CODE REVIEW

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 1]
```diff
@@ -140,6 +140,10 @@ public void testParseNoTitleStatement() throws Exception {
         metadata.setTitle(null);    
         MetadataUtils.getStatements(metadata);
         fail("This test is execpeted to throw an error");
+    }    
+    
+    public void testNoTitleWithFlag() {
+        
     }
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Just wondering: why did you commit an empty method? I see the next commit is the implementation, so why commit the empty method?

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 2]
```diff
@@ -142,9 +142,25 @@ public void testParseNoTitleStatement() throws Exception {
         fail("This test is execpeted to throw an error");
     }    
     
-    public void testNoTitleWithFlag() {
-        
-    }
+    /**
+     * Test missing rdf statement but the mandatory check is disabled,
+     * so this test is excepted to pass
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Typo on 'excepted', should be expected. But aren't all test expected to pass?

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 3]
```diff
@@ -142,9 +142,25 @@ public void testParseNoTitleStatement() throws Exception {
...
+     * 
+     * @throws Exception 
+     */
+    @Test
+    public void testParseNoTitleStatementWithFlag() throws Exception {
+        System.out.println("Test : Missing title statement");
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: sysout vs using a logger, is there any reason for choosing the former over the latter?

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 4]
```diff
@@ -142,9 +142,25 @@ public void testParseNoTitleStatement() throws Exception {
...
+        FDPMetadataParser fdpParser = new FDPMetadataParser();
+        List<Statement> stmts = ExampleFilesUtils.getFileContentAsStatements(
+                ExampleFilesUtils.FDP_METADATA_FILE, 
+                        ExampleFilesUtils.FDP_URI.toString());
+        IRI fURI = ExampleFilesUtils.FDP_URI;
+        FDPMetadata metadata = fdpParser.parse(stmts , fURI);
+        metadata.setTitle(null);    
+        List<Statement> excepted = MetadataUtils.getStatements(metadata);
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Typo again I assume, 'excepted' should be expected?

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 5]
```diff
@@ -166,6 +182,26 @@ public void testParseNoIDStatement() throws Exception {
     }
     
     /**
+     * Test missing rdf statement but the mandatory check is disabled,
+     * so this test is excepted to pass
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Typo on expected again.

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 6]
```diff
@@ -166,6 +182,26 @@ public void testParseNoIDStatement() throws Exception {
...
+    public void testParseNoIDStatementWithFlag() throws Exception {
+        System.out.println("Test : Missing ID statement");
+        FDPMetadataParser fdpParser = new FDPMetadataParser();
+        List<Statement> stmts = ExampleFilesUtils.getFileContentAsStatements(
+                ExampleFilesUtils.FDP_METADATA_FILE, 
+                        ExampleFilesUtils.FDP_URI.toString());
+        IRI fURI = ExampleFilesUtils.FDP_URI;
+        FDPMetadata metadata = fdpParser.parse(stmts , fURI);
+        metadata.setIdentifier(null); 
+        List<Statement> excepted = MetadataUtils.getStatements(metadata);
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Typo on expected again.

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 7]
```diff
@@ -305,6 +382,33 @@ public void testParseEmptyThemeTaxonomyStatement() throws Exception {
...
+        System.out.println("Test : Empty theme taxonomy statement");
+        CatalogMetadataParser parser = new CatalogMetadataParser();
+        List<Statement> stmts = ExampleFilesUtils.getFileContentAsStatements(
+                ExampleFilesUtils.CATALOG_METADATA_FILE, 
+                        ExampleFilesUtils.CATALOG_URI.toString());
+        IRI cURI = ExampleFilesUtils.CATALOG_URI;
+        CatalogMetadata metadata = parser.parse(stmts , cURI);
+        metadata.setThemeTaxonomys(new ArrayList());
+        MetadataUtils.getStatements(metadata);
+        List<Statement> excepted = MetadataUtils.getStatements(metadata);
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Why is `MetadataUtils.getStatements` called twice? (line 402 and line 403)

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 8]
```diff
@@ -305,6 +382,33 @@ public void testParseEmptyThemeTaxonomyStatement() throws Exception {
...
+        assertNotNull(excepted);
+        
+        metadata.setThemeTaxonomys(null);
+        MetadataUtils.getStatements(metadata);
+        excepted = MetadataUtils.getStatements(metadata);
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Same pattern, getStatements is called twice.

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java [hunk 9]
```diff
@@ -385,6 +539,27 @@ public void testGetDataRecordNoRMLURI() throws Exception {
...
+                        ExampleFilesUtils.DATARECORD_URI.toString());
+        IRI drURI = ExampleFilesUtils.DATARECORD_URI;
+        DataRecordMetadata metadata = parser.parse(stmts , drURI);  
+        metadata.setRmlURI(null);
+        List<Statement> excepted = MetadataUtils.getStatements(metadata);
+        assertNotNull(excepted);
+    }
+    
+    
+    /**
```
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: As a generic comment on the tests added to this class: there is a lot of repetition in each test setup. Maybe you can refactor it into a private convenience method, I can easily see about 6 lines saved for each method. With the methods added in this commit that's easily 40 lines saved.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 10]
```diff
@@ -80,10 +80,67 @@
...
      */
-    public static final int SCHEMA_DOT_ORG = 1;
+    public static final int SCHEMA_DOT_ORG_MODEL = 1;
+    
+    
+    
+    /**
+     * Convert Metadata object to RDF string for specific model
+     *
+     * @param <T>
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: You could document the template parameter as well.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 11]
```diff
@@ -80,10 +80,67 @@
...
+     * @param metadata Subclass of metadata object
+     * @param format RDF format
+     * @return RDF string
+     * @throws MetadataException
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: You could document the exception conditions as well.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 12]
```diff
@@ -112,41 +170,92 @@
...
+     * @param checkFields To indicate if mandatory fields are need to checked
+     * @return List of RDF statements
+     * @throws MetadataException
+     */
+    public static <T extends Metadata> List<Statement> getStatements(
+            @Nonnull T metadata, int metadataModel, boolean checkFields) throws 
             MetadataException {
         Preconditions.checkNotNull(metadata,
-                "Metadata object must not be null.");
+                "Metadata object must not be null.");
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Superfluous trailing whitespace.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 13]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+        }
+    }
+    
+    /**
+     * Check fdp metadata properties
+     * 
+     * @param metadata  FDPMetadata object
+     * @throws MetadataException 
+     */
+    private static void checkFDPPropeties(@Nonnull FDPMetadata metadata)
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Typo on 'properties'.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 14]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+        }
+    }
+    
+    /**
+     * Check catalog metadata properties
+     * 
+     * @param metadata  CatlogMetadata object
+     * @throws MetadataException 
+     */
+    private static void checkCatalogPropeties(@Nonnull CatalogMetadata metadata)
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Typo on 'properties'.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 15]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+     * @param metadata  FDPMetadata object
+     * @throws MetadataException 
+     */
+    private static void checkFDPPropeties(@Nonnull FDPMetadata metadata)
+            throws MetadataException {
+        try {
+            Preconditions.checkNotNull(metadata.getRepostoryIdentifier(),
+                    "Repostory ID must not be null.");
+        } catch (NullPointerException ex) {
+            throw (new MetadataException(ex.getMessage()));
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Superfluous parenthesis for the throw statement.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 16]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+     */
+    private static void checkCatalogPropeties(@Nonnull CatalogMetadata metadata)
+            throws MetadataException {
+        try {
+            Preconditions.checkNotNull(metadata.getThemeTaxonomys(),
+                    "Metadata dcat:themeTaxonomy must not be null.");
+            Preconditions.checkArgument(!metadata.getThemeTaxonomys().isEmpty(),
+                    "Metadata dcat:themeTaxonomy must not be empty.");
+        } catch (NullPointerException | IllegalArgumentException ex) {
+            throw (new MetadataException(ex.getMessage()));
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Superfluous parenthesis for the throw statement.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 17]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+     */
+    private static void checkDatasetPropeties(@Nonnull DatasetMetadata metadata)
+            throws MetadataException {
+        try {
+            Preconditions.checkNotNull(metadata.getThemes(),
+                    "Metadata dcat:theme must not be null.");
+            Preconditions.checkArgument(!metadata.getThemes().isEmpty(),
+                    "Metadata dcat:theme must not be empty.");
+        } catch (NullPointerException | IllegalArgumentException ex) {
+            throw (new MetadataException(ex.getMessage()));
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Superfluous parenthesis for the throw statement.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 18]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+        }
+    }
+    
+    /**
+     * Check dataset metadata properties
+     * 
+     * @param metadata  DatasetMetadata object
+     * @throws MetadataException 
+     */
+    private static void checkDatasetPropeties(@Nonnull DatasetMetadata metadata)
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Typo on 'properties'.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 19]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+        }
+    }
+    
+    /**
+     * Check distribution metadata properties
+     * 
+     * @param metadata  DistributionMetadata object
+     * @throws MetadataException 
+     */
+    private static void checkDistributionPropeties(@Nonnull
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Typo on 'properties'.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 20]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+     * @throws MetadataException 
+     */
+    private static void checkDistributionPropeties(@Nonnull 
+            DistributionMetadata metadata) throws MetadataException {
+        if (metadata.getAccessURL() == null
+                && metadata.getDownloadURL() == null) {
+            String errMsg
+                    = "No dcat:accessURL or dcat:downloadURL URL is provided";
+            LOGGER.error(errMsg);
+            throw (new MetadataException(errMsg));
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Superfluous parenthesis for the throw statement.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 21]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+     * @param metadata  DataRecordMetadata object 
+     * @throws MetadataException 
+     */
+    private static void checkDatarecordPropeties(@Nonnull DataRecordMetadata 
+            metadata) throws MetadataException {
+        try {
+            Preconditions.checkNotNull(metadata.getRmlURI(),
+                    "Metadata rml mapping uri must not be null.");
+        } catch (NullPointerException ex) {
+            throw (new MetadataException(ex.getMessage()));
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Superfluous parenthesis for the throw statement.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java [hunk 22]
```diff
@@ -611,6 +568,115 @@ private static void checkMandatoryProperties(Metadata metadata)
...
+        }
+    }
+    
+    /**
+     * Check datarecord metadata properties
+     * 
+     * @param metadata  DataRecordMetadata object 
+     * @throws MetadataException 
+     */
+    private static void checkDatarecordPropeties(@Nonnull DataRecordMetadata
```
Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Typo on 'properties'.

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java (hunk 3 above)
Comment (rajaram5) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Just to save some logger lines. But do you want me to use logger?

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java (hunk 5 above)
Comment (rajaram5) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: I will correct this

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java (hunk 3 above)
Comment (kburger) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Well 'want' might be too much, but if you use a logger (which will only add 3 lines: one for the logger declaration and 2 for the imports) you can at least control the output through the logging configuration. Right now the only control you have is to comment out the line. What do you think?

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java (hunk 11 above)
Comment (rajaram5) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Thanks for noticing some of this exceptions are not thrown I removed it.

File: src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java (hunk 3 above)
Comment (rajaram5) [inline src/test/java/nl/dtl/fairmetadata4j/utils/MetadataUtilsTest.java]: Yes, I agree I will replace them with logger

**Comment by kburger** [comment]
Some minor codestyle issues remain, but we'll discuss that at a later point. I'm good with these changes.

File: src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java (hunk 20 above)
Comment (rajaram5) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: But the fix for this will add one more line to this section right?

Comment (kburger) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: What is meant was:
```java
throw new MetadataException(errMsg);
```

vs

```java
throw (new MetadataException(errMsg));
```

Comment (rajaram5) [inline src/main/java/nl/dtl/fairmetadata4j/utils/MetadataUtils.java]: Ah I see. I will fix it

**Pull request merged**